clients/src/main/java/org/apache/kafka/common/protocol/Message.java

```diff
@@ -47,7 +47,20 @@
      *                      If the specified version is too new to be supported
      *                      by this software.
      */
-    int size(ObjectSerializationCache cache, short version);
+    default int size(ObjectSerializationCache cache, short version) {
+        MessageSizeAccumulator size = new MessageSizeAccumulator();
+        addSize(size, cache, version);
+        return size.totalSize();
+    }
+
+    /**
+     * Add the size of this message to an accumulator.
+     *
+     * @param size          The size accumulator to add to
+     * @param cache         The serialization size cache to populate.
+     * @param version       The version to use.
+     */
+    void addSize(MessageSizeAccumulator size, ObjectSerializationCache cache, short version);
```

Comment:
I'm thinking about how to simplify this process.

Could we reuse the method ```void write(Writable writable, ObjectSerializationCache cache, short version)``` ? Maybe we can create a ```Writable``` instance but it does not write data to any output. Instead, it calculate the size of buffer according to input data.